golang/golang_chapter_0050/text.md

```diff
@@ -0,0 +1,233 @@
+
+```golang {.task_source #golang_chapter_0050_task_0010}
+package main
+import "fmt"
+
+func count(){
+    // ваш код здесь 
+}
+
+func main() {
+	s := "Язык Go является регистрозависимым языком"
+	fmt.Println(count(s))
+}
+```  
+
+Воспользуйтесь функцией `strings.Split`. {.task_hint}
+
+```golang {.task_answer}
+package main
+
+import (
+	"fmt"
+	"strings"
+)
+
+func count(s string) int {
+	res := strings.Split(s, " ")
+	return len(res)
+}
+
+func main() {
+	s := "Язык Go является регистрозависимым языком"
+
+	fmt.Println(count(s))
+}
+```
+
+В переменной `slice` типа `[]uint` содержится некоторый срез. Реализуйте тело функции `maxEl`, которая принимает на вход срез данного типа и возвращает максимальный элемент. {.task_text}
+
+```golang {.task_source #golang_chapter_0050_task_0020}
+package main
+import "fmt"
+
+func maxEl(slice []uint) uint{
+    // ваш код здесь 
+}
+
+func main() {
+	slice := []uint{9, 10, 4, 2}
+
+	fmt.Println(maxEl(slice))
+}
+```  
+
+Объявите переменную типа `uint`. По умолчанию в ней лежит значение `0`. Это минимальное из возможных значений для переменной данного типа. Реализуйте цикл по всем элементам среза. Если внутри цикла найдется элемент больше, чем значение данной переменной, присвойте переменной это значение. По окончанию цикла верните значений переменной. {.task_hint}
+
+```golang {.task_answer}
+package main
+
+import (
+	"fmt"
+)
+
+func maxEl(slice []uint) uint {
+	var res uint
+
+	for _, el := range slice {
+		if el > res {
+			res = el
+		}
+	}
+
+	return res
+}
+
+func main() {
+	slice := []uint{9, 10, 4, 2}
+
+	fmt.Println(maxEl(slice))
+
+}
+```
+
+В переменной `slice` типа `[]uint16` содержится некоторый срез. Реализуйте тело функции `diff`, которая принимает на вход срез данного типа и возвращает разницу между максимальным и минимальным элементами среза. {.task_text}
+
+```golang {.task_source #golang_chapter_0050_task_0030}
+package main
+import "fmt"
+
+func diff(slice []uint16) uint16 {
+    // ваш код здесь 
+}
+
+func main() {
+	slice := []uint16{9, 10, 4, 2}
+
+	fmt.Println(diff(slice))
+
+}
+```  
+
+Максимальное значение для переменной типа `uint16` равно `65535`. {.task_hint}
+
+```golang {.task_answer}
+package main
+
+import (
+	"fmt"
+)
+
+func diff(slice []uint16) uint16 {
+	var maxEl uint16
+	var minEl uint16 = 65535
+
+	for _, el := range slice {
+		if el > maxEl {
+			maxEl = el
+		}
+		if el < minEl {
+			minEl = el
+		}
+	}
+
+	return maxEl - minEl
+}
+
+func main() {
+	slice := []uint16{9, 10, 4, 2}
+
+	fmt.Println(diff(slice))
+
+}
+
+```
+
+## Резюме 
+1. В Go используются срезы, которые реализуют массивы переменной длины. Срезы могут быть как одномерными, так и многомерными. 
+2. Срез представляет собой указатель, длину и емкость.
+3. Срез всегда ссылается на базовый массив. 

```

Comment:
Мне кажется, в главе нужно рассказать про довольно типичные грабли при работе со слайсами.

> re-slicing a slice doesn’t make a copy of the underlying array. The full array will be kept in memory until it is no longer referenced. Occasionally this can cause the program to hold all the data in memory when only a small piece of it is needed.

https://go.dev/blog/slices-intro